kiwi/builder/live.py

```diff
@@ -230,65 +230,91 @@ def create(self) -> Result:
         }
         filesystem_setup = FileSystemSetup(
             self.xml_state, self.root_dir
         )
-        root_image = Temporary().new_file()
-        with LoopDevice(
-            root_image.name,
-            filesystem_setup.get_size_mbytes(root_filesystem),
-            self.xml_state.build_type.get_target_blocksize()
-        ) as loop_provider:
-            loop_provider.create()
+        if root_filesystem != 'squashfs':
+            # Create a filesystem image of the specified type
+            # and put it into a SquashFS container
+            root_image = Temporary().new_file()
+            with LoopDevice(
+                root_image.name,
+                filesystem_setup.get_size_mbytes(root_filesystem),
+                self.xml_state.build_type.get_target_blocksize()
+            ) as loop_provider:
+                loop_provider.create()
+                with FileSystem.new(
+                    name=root_filesystem,
+                    device_provider=loop_provider,
+                    root_dir=self.root_dir + os.sep,
+                    custom_args=filesystem_custom_parameters
+                ) as live_filesystem:
+                    live_filesystem.create_on_device()
+                    log.info(
+                        '--> Syncing data to {0} root image'.format(root_filesystem)
+                    )
+                    live_filesystem.sync_data(
+                        Defaults.
+                        get_exclude_list_for_root_data_sync() + Defaults.
+                        get_exclude_list_from_custom_exclude_files(self.root_dir)
+                    )
+
+            log.info('--> Creating squashfs container for root image')
+            self.live_container_dir = Temporary(
+                prefix='live-container.', path=self.target_dir
+            ).new_dir()
+            Path.create(self.live_container_dir.name + '/LiveOS')
+            shutil.copy(
+                root_image.name, self.live_container_dir.name + '/LiveOS/rootfs.img'
+            )
+            with FileSystem.new(
+                name='squashfs',
+                device_provider=DeviceProvider(),
+                root_dir=self.live_container_dir.name,
+                custom_args={
+                    'compression':
+                        self.xml_state.build_type.get_squashfscompression()
+                }
+            ) as live_container_image:
+                container_image = Temporary().new_file()
+                live_container_image.create_on_file(
+                    container_image.name
+                )
+                Path.create(self.media_dir.name + '/LiveOS')
+                os.chmod(container_image.name, 0o644)
+                shutil.copy(
+                    container_image.name,
+                    self.media_dir.name + '/LiveOS/squashfs.img'
+                )
+        else:
+            # Put the root filesystem into SquashFS directly
```

Comment:
I like the way you added the feature in code but I think it's confusing that the filesystem attribute changes the layout of the live ISO. There is the ```flags=overlay|dmsquash``` attribute which we used to control the layout of the live ISO and there was a reason why ```filesystem``` was forbidden because you actually have no choice for the filesystem, it will always be some squashfs+overlay layout. So if you don't mind please use the value for the flags to differentiate the layout and not filesystem. You have the current value available in the variable ```self.live_type```. 

Feel free to run your new layout with ```if self.live_type == 'overlay':```.

Reply:
That way you also don't have to change the schema and the only missing part is then to double check if the documentation still matches the code for flags="overlay"

Reply:
If you don't like this you can also add another flag value but please don't let us abuse the filesystem attribute :)

Reply:
`overlay` vs. `dmsquash` is independent of the actual file system, it's just how writing and persistence are implemented. With `overlay`, the overlay is on top of the filesystem, while with `dmsquash` the overlay is on the block level below the filesystem. So it's perfectly fine to use `flags="overlay" filesystem="xfs"` or `flags="dmsquash" filesystem="btrfs"`.

IMO using `filesystem` for this is intuitive, because that is what the running live system will see as effective file system for `/`.

Even the `flags="dmsquash" filesystem="squashfs"` combination should work according to the [dracut.cmdline man page](https://manpages.opensuse.org/Tumbleweed/dracut/dracut.cmdline.7.en.html#Booting_live_images):

```
 For OverlayFS mount overlays, the filesystem structure may also be a direct compression of the root filesystem:

     squashfs.img          |  SquashFS from LiveCD .iso
        !(mount)
        /bin               |  Live filesystem
        /boot              |
        /dev               |
        ...                |
```

I didn't test that yet, but I'll try to do a build.

Reply:
> IMO using filesystem for this is intuitive, because that is what the running live system will see as effective file system 
for /

This is true but it will always be a squashfs. If you say ```flags="dmsquash" filesystem="ext3"``` it will be accepted and ignored. Why not just add a new flag e.g

```xml
<type image="iso" ... flags="direct-squashfs"/>
```
or any name that makes it clear.

Reply:
> the overlay is on the block level below the filesystem. So it's perfectly fine to use flags="overlay" filesystem="xfs"

This is also true but I don't think we want to support that now or in the future. The squashfs embedded filesystem image was a layout only introduced to stay compatible with the old concept which uses only the device-mapper. I don't see any benefit to place an XX filesystem image inside of the squashfs. That's also the reason why I really like your change, it gets rid of this unneeded indirection. So why add code that allows for a setting we don't want to support ?

Reply:
ok, I was wrong when I wrote the filesystem setting got ignored, sorry I did no read carefully enough. But the question stays, why do we want to support this ? Does it make sense ?

Reply:
ok, I agree so let's keep it that way. Can you add a few words in the documentation about the change ? I think this chapter here would be a good place: https://osinside.github.io/kiwi/building_images/build_live_iso.html
Thanks